design/internal-redirect_design.md

```diff
@@ -0,0 +1,170 @@
+    // An empty list means all routes are allowed.
+    // +optional
+    AllowedRouteNames []string `json:"allowedRouteNames,omitempty"`
+}
+
+```
+
+Update `Route` accordingly. 
+The `AllowedRouteNames` predicates declares a list of allowed route's name, so a `Name` field is required on `Route` to be able to use it.
+
+```Go
+type Route struct {
+    [... other members ...]
+
+    // Name of the route.
+    // +optional
+    Name string `json:"name,omitempty"`
+
+    // InternalRedirectPolicy defines when to handle redirect responses internally.
+    // +optional
+    InternalRedirectPolicy *InternalRedirectPolicy `json:"internalRedirectPolicy,omitempty"`
+}
+```
+
+The common DAG struct will be updated accordingly:
+
+```Go
+
+// InternalRedirectPolicy defines if envoy should handle redirect response internally instead of sending it downstream.
+// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-internalredirectpolicy
+type InternalRedirectPolicy struct {
+    // MaxInternalRedirects An internal redirect is not handled, unless the number of previous internal redirects that a downstream request has encountered is lower than this value
+    MaxInternalRedirects uint32
+
+    // RedirectResponseCodes If unspecified, only 302 will be treated as internal redirect.
+    // Only 301, 302, 303, 307 and 308 are valid values
+    RedirectResponseCodes []uint32
+
+    // AllowCrossSchemeRedirect Allow internal redirect to follow a target URI with a different scheme than the value of x-forwarded-proto.
+    AllowCrossSchemeRedirect string
+
+    DenyRepeatedRouteRedirect bool
+
+    // AllowListedRoutesPredicate accepts only explicitly allowed target routes.
+    AllowedRouteNames []string
+}
+
+type Route struct {
+    [... other members ...]
+
+    // InternalRedirectPolicy defines if envoy should handle redirect response internally instead of sending it downstream.
+    InternalRedirectPolicy *InternalRedirectPolicy
+}
+```
+
+### Enabling internal redirect in Envoy
+In order to enable internal redirect in Envoy we will update `contour/internal/envoy/route.go` and map the values from DAG's route to [protobuf](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-internal-redirect-policy).
+
+
+## Alternatives Considered
+
+N/A
+
+## Security Considerations
+
+When targeting an untrusted backend which can return any value in the `Location` header, the gateway operator may want to be able to limit the scope of the internal redirect to a list of predefined routes by populating the `allowedRouteNames` field.
+But in the case of multitenant Contour deployment, 2 teams may declare routes using the same name, defeating the purpose of the allowed routes list.
+
+To prevent unexpected redirect behavior in that case, Contour should ensure route name uniqueness.
+
+## Compatibility
```

Comment:
might be worth considering if there are other *Policy fields that this is not valid to or doesn't make sense to configure with on a HTTPProxy Route, e.g. RequestRedirectPolicy or DirectResponsePolicy

Reply:
At the Envoy level, `Route`, `RequestRedirect` and `DirectResponse` are 3 mutually exclusive members and `InternalRedirect` is a member of `Route`.

It should behave like any other member of `HTTPProxy.Route` that is ignored if `RequestRedirectPolicy` or `DirectResponsePolicy` policy is defined (`TimeoutPolicy`, `RetryPolicy`, …).